examples/knative/powershell/kn-ps-webhook/handler.ps1

```diff
@@ -0,0 +1,136 @@
+   param()
+   Write-Host "$(Get-Date) - Processing Init`n"
+
+   Write-Host "$(Get-Date) - Init Processing Completed`n"
+}
+
+Function Process-Shutdown {
+   [CmdletBinding()]
+   param()
+   Write-Host "$(Get-Date) - Processing Shutdown`n"
+
+   Write-Host "$(Get-Date) - Shutdown Processing Completed`n"
+}
+
+Function Process-Handler {
+   [CmdletBinding()]
+   param(
+      [Parameter(Position=0,Mandatory=$true)]$Headers,
+      [Parameter(Position=1,Mandatory=$true)]$Body
+   )
+
+   try {
+      $jsonSecrets = ${env:WEBHOOK_SECRET} | ConvertFrom-Json
+   } catch {
+      throw "`nK8s secrets `$env:WEBHOOK_SECRET does not look to be defined"
+   }
+
+   Write-host "$(Get-Date) - Processing Body ...`n"
+   try {
+      if ($Body -is [string]) {
+         $Body = [System.Text.Encoding]::UTF8.GetBytes($Body)
+      }
+   } catch {
+      throw "`nFailed to encode `$Body input - `$Error[0]"
+   }
+
+   if(${env:SERVICE_DEBUG} -eq "true") {
+      Write-Host "$(Get-Date) - [Function Secrets]:`n${env:WEBHOOK_SECRET}`n"
+   }
+
+   # Check to see if basic auth has been configured for webhook
+   if(${jsonSecrets}.WEBHOOK_USERNAME -ne $NULL -and ${jsonSecrets}.WEBHOOK_PASSWORD -ne $NULL) {
+      Write-host "$(Get-Date) - Checking for authorization ...`n"
+
+      # Expected base64 value for the configured webhook username/pass
+      $expectedPair = "$(${jsonSecrets}.WEBHOOK_USERNAME):$(${jsonSecrets}.WEBHOOK_PASSWORD)"
+      $expectedBytes = [System.Text.Encoding]::ASCII.GetBytes($expectedPair)
+      $expectedBase64 = "Basic " + [System.Convert]::ToBase64String($expectedBytes)
+
+      if(${env:SERVICE_DEBUG} -eq "true") {
+         Write-Host "$(Get-Date) - [Expected Authorization Header]:`n$expectedBase64`n"
+      }
+
+      # Check for the supplied webhook username/pass
+      if(${Headers}.Authorization) {
+         if($(${Headers}.Authorization) -ne "$expectedBase64") {
+            Write-Error "$(Get-Date) - Invalid Authorization`n"
+            $exception = New-Object Exception
+            $httpException = New-Object System.Net.Http.HttpRequestException -ArgumentList "Invalid Authorization",$exception,400
+            throw $httpException
+         }
+      } else {
+         Write-Error "$(Get-Date) - Authorization header was not provided`n"
+         $exception = New-Object Exception
+         $httpException = New-Object System.Net.Http.HttpRequestException -ArgumentList "Authorization header was not provided",$exception,400
+         throw $httpException
+      }
+   }
+
+   # Retrieve webhook sink URL
+   $webhookSinkUrl = $(${jsonSecrets}.WEBHOOK_SINK_URL)
+
+   # Retrieve the CloudEvent Type
+   $cloudEventType = $(${jsonSecrets}.WEBHOOK_CE_EVENT_TYPE)
+
+   # Retrieve forwarded IP[0] (client address), if that is empty, set to function hostname
+   try {
+      $cloudEventSource = ((${headers}.'X-Forwarded-For') -split ',')[0]
+
+      if($cloudEventSource -eq "") {
+         $cloudEventSource = [System.Net.Dns]::GetHostName()
```

Comment:
What is this doing then?

Reply:
This defaults to function hostname IF X-Forwarded-For header is ever blank. Its more of a safety, as I expect it to also contain something :)

Reply:
Sorry for not being clear, I was wondering why you have 85 bc it's also the same line as in 89?